byzcoin/struct.go

```diff
@@ -639,6 +639,14 @@ func (bc *bcNotifications) unregisterForBlocks(ch waitChannel) {
 
 	for i, listener := range bc.blockListeners {
 		if listener == ch {
+		emptyChannel:
```

Comment:
It doesn't really drain a channel, only the available messages. If the channels are supposed to be closed at some point, you can use `for range ch {}`, if not, then, that's harder, there isn't a correct way to do it (it comes down to the Turing halting problem). Maybe add a `close` before breaking to catch channel misuse?

Reply:
True - there might be a block coming in just now. So I'll just close it here and handle that case where it got stuck before:

https://github.com/dedis/cothority/blob/8eec9f1265ba03060878d92f2d88e7dac421de6a/byzcoin/struct.go#L621

Reply:
In fact, it might be enough. The `bc.Lock` _nearly_ assures that no new messages get in the channel. 

The _nearly_ being because of this:
```
// Pseudocode
bc.Lock()
go func(){
  bc.blockListeners[0] <- something
}
bc.UnLock()
```

So there is a small time-window between the `bc.Lock` and the execution of the go-routine, where a new message might come in. But as the main purpose of the drain is to stop the go-routine to be stuck, and the `bc.blockListeners` is of length 1, it would have to race twice the `bc.Lock` and `go func()` before the go-routine gets stuck anyway.

But if you have a cleaner solution, I'm all for it...

Reply:
I could make the wait-channel of length 10 in line 630, and then replace the go-routine in line 620 with a `select` and print a warning if the channel overflows - would that be better?

Reply:
Hum, I'm more versed in the second option, that allow for most of the gouroutine to hopefully finish normally.

Reply:
OK, added a queue that will throw away blocks if they are not fetched in time.